src/fsharp/FSharp.Core/iseq.fs

```diff
@@ -0,0 +1,1345 @@
+                        override this.ProcessNext (input:'T) : bool =
+                            this.State.Value.Add input && TailCall.avoid (next.ProcessNext input) }}
+
+        [<CompiledName "Exists">]
+        let inline exists f (source:ISeq<'T>) =
+            source.Fold (fun pipeIdx ->
+                { new Folder<'T, bool> (false) with
+                    override this.ProcessNext value =
+                        if f value then
+                            this.Result <- true
+                            (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        Unchecked.defaultof<_> (* return value unused in Fold context *) })
+
+        [<CompiledName "Exists2">]
+        let inline exists2 (predicate:'T->'U->bool) (source1:ISeq<'T>) (source2: ISeq<'U>) : bool =
+            source1.Fold (fun pipeIdx ->
+                upcast { new FolderWithPostProcessing<'T,bool,IEnumerator<'U>>(false,source2.GetEnumerator()) with
+                    override this.ProcessNext value =
+                        if this.State.MoveNext() then
+                            if predicate value this.State.Current then
+                                this.Result <- true
+                                (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        else
+                            (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        Unchecked.defaultof<_> (* return value unused in Fold context *)
+
+                    override this.OnComplete _ = ()
+                    override this.OnDispose () = this.State.Dispose () })
+
+        [<CompiledName "Contains">]
+        let inline contains element (source:ISeq<'T>) =
+            source.Fold (fun pipeIdx ->
+                { new Folder<'T, bool> (false) with
+                    override this.ProcessNext value =
+                        if element = value then
+                            this.Result <- true
+                            (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        Unchecked.defaultof<_> (* return value unused in Fold context *) })
+
+        [<CompiledName "ForAll">]
+        let inline forall predicate (source:ISeq<'T>) =
+            source.Fold (fun pipeIdx ->
+                { new Folder<'T, bool> (true) with
+                    override this.ProcessNext value =
+                        if not (predicate value) then
+                            this.Result <- false
+                            (this :> IOutOfBand).StopFurtherProcessing pipeIdx
+                        Unchecked.defaultof<_> (* return value unused in Fold context *) })
+
+        [<CompiledName "ForAll2">]
+        let inline forall2 predicate (source1:ISeq<'T>) (source2:ISeq<'U>) : bool =
+            source1.Fold (fun pipeIdx ->
+                upcast { new FolderWithPostProcessing<'T,bool,IEnumerator<'U>>(true,source2.GetEnumerator()) with
```

Comment:
Stylistically I think I would prefer a tailing ``:> _`` than an explicit leading ``upcast`` in all such cases